### PR TITLE
Support for SPS-Execution-Context Custom Header

### PR DESCRIPTION
**Pull request by travisgosselin**
SPS Commerce will need to support a non-invasive method of applying customer / dataflow context to services. This is used internally by support to interact with services during custom configuration modifications, and eventually useable by external customers to test their own variations in non-production data contexts at times.

### CODE REVIEW

File: standards/request-response.md [hunk 1]
```diff
@@ -493,6 +493,36 @@ The usage of non-standard headers is not considered custom headers. For example,
...
+
+- A request without the presence of this header **MUST** indicate that the execution context is under the `production` dataflow and customer configuration, if applicable.
+- The header value contains any dynamic __string__ value representing a named configuration or tagged dataflow preset.
+    - __preprod__ - Well-known value used to represent non-production customer configuration where named configuration is not supported. Typically used for legacy support. Usage of dynamically named configuration sets is preferred.
+- An invalid or unsupported header value provided **MUST** result in a `400 - Bad Request` following standard [error format](errors.md#400-bad-request).
+- The header value **MUST NOT** exceed a maximum length of 100 characters.
+- The header value **SHOULD** contain human-readable tag for the context.
+- The header **MUST NOT** be used in a response.
```
Comment (alexander-ivakhnenko) [inline standards/request-response.md:510]: Why? It might be useful to return back (especially if it's async) to caller so the integrity of the context is maintained. That said, it should be preserved in response only if it was as part of the original request.

File: standards/request-response.md [hunk 2]
```diff
@@ -493,6 +493,36 @@ The usage of non-standard headers is not considered custom headers. For example,
...
+
+**Type**: Request
+
+**Support**: OPTIONAL
+
+**Description**: The Execution Context header provides a standard method of indicating the dataflow context that a request should be processed with. This is used to differentiate or test changes in customer configuration of non-production data. This isolates the version of the configuration state used in a request and allows identifying transactions that should leverage a specific tagged or aliased configuration set.
+
+- A request without the presence of this header **MUST** indicate that the execution context is under the `production` dataflow and customer configuration, if applicable.
+- The header value contains any dynamic __string__ value representing a named configuration or tagged dataflow preset.
+    - __preprod__ - Well-known value used to represent non-production customer configuration where named configuration is not supported. Typically used for legacy support. Usage of dynamically named configuration sets is preferred.
```
Comment (alexander-ivakhnenko) [inline standards/request-response.md:506]: "where named configuration is not supported".. maybe rephrase this a bit. Should say something like "preprod is a well-known value used to represent non-production customer configuration and can be supported as static mode in some legacy services without full dynamic support. Usage of dynamically named configuration sets is preferred.

File: standards/request-response.md (hunk 1 above)
Comment (travisgosselin) [reply standards/request-response.md:510]: I agree with this - in fact we had this discussion and you wanted as request only. But yes if we do modify this:
- What is the response if none in the request?
- Is this response header always given? (I think it should be)

File: standards/request-response.md [hunk 3]
```diff
@@ -493,6 +493,36 @@ The usage of non-standard headers is not considered custom headers. For example,
...
 
+#### SPS-Execution-Context
+
+**Type**: Request
+
+**Support**: OPTIONAL
+
+**Description**: The Execution Context header provides a standard method of indicating the dataflow context that a request should be processed with. This is used to differentiate or test changes in customer configuration of non-production data. This isolates the version of the configuration state used in a request and allows identifying transactions that should leverage a specific tagged or aliased configuration set.
+
+- A request without the presence of this header **MUST** indicate that the execution context is under the `production` dataflow and customer configuration, if applicable.
```
[the remark this answers is not in the document]
Comment (travisgosselin) [reply standards/request-response.md:504]: I couldn't help but think that, this concept really doesn't apply to EVERY API ever made?

File: standards/request-response.md (hunk 2 above)
Comment (travisgosselin) [inline standards/request-response.md:506]:
```suggestion
    - __preprod__ - Well-known value used to represent non-production customer configuration and can be supported as a static mode in some legacy services without full dynamic support. Usage of dynamically named configuration sets is preferred.
```

File: standards/request-response.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexander-ivakhnenko) [reply standards/request-response.md:504]: ah, in that context. No, some APIs don't need it. Maybe we should separate the fact of APIs that by concept support it but don't have this header and APIs that are not part of transactional processing (or don't have to support context) and that's where it will be non-applicable case.

File: standards/request-response.md (hunk 1 above)
Comment (alexander-ivakhnenko) [reply standards/request-response.md:510]: What is the response if none in the request?... I'd say optional unless it's by-design of this API to return it. Maybe in that situation, it's not even a header, but a response body. Like API that returns available contexts.

Is this response header always given? (I think it should be) ... as long as it's provided in the request, otherwise, it's optional to return based on the behavior of API. (if I understood question right)

File: standards/request-response.md [hunk 4]
```diff
@@ -493,6 +493,36 @@ The usage of non-standard headers is not considered custom headers. For example,
...
+- The header value contains any dynamic __string__ value representing a named configuration or tagged dataflow preset.
+    - __preprod__ - Well-known value used to represent non-production customer configuration and can be supported as a static mode in some legacy services without full dynamic support. Usage of dynamically named configuration sets is preferred.
+- An invalid or unsupported header value provided **MUST** result in a `400 - Bad Request` following standard [error format](errors.md#400-bad-request).
+- The header value **MUST NOT** exceed a maximum length of 100 characters.
+- The header value **SHOULD** contain human-readable tag for the context.
+- The header **MUST NOT** be used in a response.
+- The header **SHOULD** be propagated to any outgoing requests to retain the context for downstream usage.
```
Comment (nickclarity) [inline standards/request-response.md:511]: Do we need to add language regarding services that support both `preprod` and `prod` in the same context?  An example would be DEX supporting the `preprod` context which depends on Rules management that makes no distinction between `preprod` and `prod`.  Obviously we wouldn't expect Rules Management to return a 400.

Comment (nickclarity) [reply standards/request-response.md:511]: Or should we leave it up to the individual API's documentation to make it clear there is no distinction.

File: standards/request-response.md (hunk 1 above)
Comment (travisgosselin) [reply standards/request-response.md:510]: So to clarify: The header SHOULD be returned as a response header if it was provided in the request?

- Why not make it a "MUST BE RETURNED" if in the request for consistency?
- Is there a scenario where a service may want to return the response header for every request, regardless of if its present in the request? If so, does that imply we need a value for "PRODUCTION" to rally around (like "PREPROD")?

File: standards/request-response.md (hunk 4 above)
Comment (travisgosselin) [reply standards/request-response.md:511]: If I understand correctly, this issue is because your jumping ahead to "phase 2" of this transition where you wouldn't use "preprod", it uses dynamic rules and overrides. Take a look at internal docs about preprod context to see if that makes sense...

File: standards/request-response.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexander-ivakhnenko) [reply standards/request-response.md:504]: I maybe read too deep into it - "if applicable" in that sentence corresponds to the behavior of using "prod" when no header is provided as it reads. There is no API mentioned in the sentence itself, and people might try to apply "if applicable" to described behavior and not to the fact that API does not support context by design. Maybe just clarifying like "if applicable to this API" will be sufficient.

File: standards/request-response.md (hunk 1 above)
Comment (alexander-ivakhnenko) [reply standards/request-response.md:510]: 1 - makes sense to me
2 - that's something we touched on earlier and likely a more long-term view of this. PRODUCTION essentially can be equal to EMPTY. Might be something worth discussing a bit more, but I don't see reason why not.

File: standards/request-response.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (travisgosselin) [reply standards/request-response.md:504]: ah gotcha! I understand now! Make sense.

File: standards/request-response.md (hunk 1 above)
Comment (travisgosselin) [reply standards/request-response.md:510]: Well I think it would be odd to have a header with no value.

So what we are saying is if the header is supplied on the request, then the response returns it. If it is not supplied, it does not return it. Based on this, you would never expect to see the request or response header provided that included the term (or similar term) "production"?

File: standards/request-response.md (hunk 3 above)
Comment (travisgosselin) [inline standards/request-response.md:504]:
```suggestion
- A request without the presence of this header **MUST** indicate that the execution context is under the `production` dataflow and customer configuration, if applicable to this API.
```

File: standards/request-response.md [hunk 5]
```diff
@@ -493,6 +493,36 @@ The usage of non-standard headers is not considered custom headers. For example,
...
+
+- A request without the presence of this header **MUST** indicate that the execution context is under the `production` dataflow and customer configuration, if applicable to this API.
+- The header value contains any dynamic __string__ value representing a named configuration or tagged dataflow preset.
+    - __preprod__ - Well-known value used to represent non-production customer configuration and can be supported as a static mode in some legacy services without full dynamic support. Usage of dynamically named configuration sets is preferred.
+- An invalid or unsupported header value provided **MUST** result in a `400 - Bad Request` following standard [error format](errors.md#400-bad-request).
+- The header value **MUST NOT** exceed a maximum length of 100 characters.
+- The header value **SHOULD** contain human-readable tag for the context.
```
Comment (travisgosselin) [inline standards/request-response.md:509]: Is this generally true? Might be different types of "integers" that get passed today, like rules management ids.... I think generally its important to move to immutable strings here... maybe not "human readable" though??? Should indicate that these are opaque string values.

**Review by acchristensen — APPROVED** [review]
👍 👍 👍

**Review by alexander-ivakhnenko — APPROVED** [review]
